app/assets/javascripts/password_validation.js

```diff
@@ -1,43 +1,78 @@
-$("#user_password").on("ajax:success", (event, xhr) => {
-  const { score } = xhr;
-  const suggestions = xhr.suggestions.length
-    ? xhr.suggestions
-    : "Choose a password that is harder to guess";
+const getSuggestionHtml = suggestions => {
+  const headingHtml = `
+    <span class="govuk-error-message govuk-!-margin-bottom-1">
+        Choose a password that is harder to guess${suggestions.length ? ':' : ''}
+    </span>
+  `;
+  const suggestionListHtml = `
+    <ul class="govuk-error-message govuk-list govuk-list--bullet">
+        ${suggestions.map(s => `<li>${s}</li>`).join('')}
+    </ul>
+  `;
+  return suggestions.length ? `${headingHtml}${suggestionListHtml}` : headingHtml;
+};
 
+const getScoreHtml = score => {
+  const scoreClassification = score < 2 ? 'weak' : (score < 4 ? 'medium' : 'good');
+  return `
+    Password strength:
+    <span class="password-strength-score score-${scoreClassification}">${scoreClassification}</span>
+  `;
+};
+
+const updateScore = score => $('.password-strength').html(getScoreHtml(score));
+
+const updateError = suggestions => {
+  $('#password').addClass('govuk-form-group--error');
+  $('#password')
+    .find('input')
+    .addClass('govuk-input--error');
+  $('.password-error-message')
+    .css('display', 'inline-block')
+    .html(`<span class="govuk-visually-hidden">Error:</span> ${getSuggestionHtml(suggestions)}`);
+};
+
+const removeError = () => {
+  $('#password').removeClass('govuk-form-group--error');
+  $('#password')
+    .find('input')
+    .removeClass('govuk-input--error');
+  $('.govuk-error-message').css('display', 'none');
+};
+
+$('#user_password').on('ajax:success', event => {
+  const detail = event.detail;
+  const [data, status] = detail;
+  if (!data || typeof data !== 'object' || status !== 'OK') return;
+  const { score, suggestions } = data;
+  if (typeof score !== 'number') return;
```

Comment:
how would score not be a number?

Reply:
Was being a bit aggressive with my duck typing, to intentionally make it clear that it needs to be a number, and keep us safe if ever the score changed to some other format (upstream or unintentionally us) which could make password scoring work in strange ways.

It will also make it safer to handle unexpected things like score being `undefined` or `null` for any number of reasons, because things like `undefined < 4` is `false` and `null < 4` is `true` which I'd rather not deal with.